### PR TITLE
Load apis per app

### PR DESCRIPTION
**Pull request by alfetopito**
# Summary

Split API loading per APP.

Right now there are only `gp-v1` APIs, so all apps are empty.

# Notes

Not very happy with this either.
I would like to conditionally load APIs based on the `APP` env var.
But that might cause the APIs to be potentialy `undefined`, which will make the refactor a nightmare.

Any better ideas?

### CODE REVIEW

**Comment by W3stside** [comment]
Screenshot from 2021-01-21 19-26-30
:scream: :skull:  :skull: RIP

File: src/api/wallet/composeProvider.ts [hunk 1]
```diff
@@ -10,7 +10,7 @@ import { TransactionConfig } from 'web3-core'
 import { numberToHex, hexToNumber } from 'web3-utils'
 import { isWalletConnectProvider, Provider } from './providerUtils'
 import { logDebug } from 'utils'
-import { web3 } from 'api'
+import { web3 } from 'api/gpV1'
```
Comment (W3stside) [inline src/api/wallet/composeProvider.ts:13]: would be cool to have one central web3 api

File: src/apps/gp-v1/pages/Settings.tsx [hunk 2]
```diff
@@ -3,7 +3,7 @@ import { useForm, Resolver as FormResolver, FieldErrors, ResolverResult } from '
 import styled from 'styled-components'
 import { DevTool } from 'HookFormDevtool'
 
-import { walletApi } from 'api'
+import { walletApi } from 'api/gpV1'
```
Comment (W3stside) [inline src/apps/gp-v1/pages/Settings.tsx:6]: and central walletapi also would be kewl

**Comment by W3stside** [comment]
i dont hate it

**Comment by W3stside** [comment]
but

**Comment by W3stside** [comment]
i dont love it neither

File: src/api/wallet/composeProvider.ts (hunk 1 above)
Comment (alfetopito) [reply src/api/wallet/composeProvider.ts:13]: Well, every app can export their own web instance exactly the same way as `gp-v1` does.

But I get your point. Every app will need `web3`. Will change that.

File: src/apps/gp-v1/pages/Settings.tsx (hunk 2 above)
Comment (alfetopito) [reply src/apps/gp-v1/pages/Settings.tsx:6]: Same point as for web3, except, will for instance the explorer need a wallet api?

**Comment by alfetopito** [comment]
> i dont love it neither

Same.
Any other suggestions?

**Comment by W3stside** [comment]
Apart from creating a gp-v1-api repo and importing those there, nope. I was
trolling low key before, this is good

File: src/PrivateRoute.tsx [hunk 3]
```diff
@@ -12,7 +12,7 @@ const useIsWalletConnected = (): {
   useEffect(() => {
     import(
       /* webpackChunkName: "API_chunk"*/
-      'api'
+      'apps/gp-v1/api'
     )
       .then(({ walletApi }) => walletApi.isConnected())
       .then((isConnected) => {
```
Comment (alfetopito) [inline src/PrivateRoute.tsx:18]: Next refactor step would be to split the hooks between shared/common and app specific.

Then we can import common from base /api instances and so on.

File: src/api/index.ts [hunk 4]
```diff
@@ -181,17 +181,5 @@ function createTcrApi(web3: Web3): TcrApi | undefined {
   return tcrApi
 }
 
-// Build APIs
```
Comment (anxolin) [inline src/api/index.ts:184]: good call on not building all the APIs

File: src/apps/gp-v1/api.ts [hunk 5]
```diff
@@ -0,0 +1,39 @@
+import {
```
Comment (anxolin) [inline src/apps/gp-v1/api.ts:1]: I like that you moved this to the app! good call

**Review by anxolin — APPROVED** [review]
The PR looks good to me! I think u did the right call moving things to their folder.

But for the new APPs, altough I see we can continue the pattern of using folders in api/<integration> , I think we can explore the approach of using plain old functions instead of the Class approach we took last time. 

Like https://github.com/gnosis/gp-swap-ui/blob/develop/src/custom/utils/operator.ts
Maybe I can even make a PR with the suggestion.

Using a functional approach is nicer for tree-shaking, specially the APIs have a lot of method, and I think at some points it got a bit too much (too big classes with hierarchies, etc).

By the way, we can bring this file

WDYT?

**Comment by anxolin** [comment]
> Not very happy with this either.
It's fine :) 

>I would like to conditionally load APIs based on the APP env var. But that might cause the APIs to be potentially undefined, which will make the refactor a nightmare.

I don't think we should do this. Every app pulls the APIs he wants. Actually, I'm suggesting pooling the methods directly instead of using classes. this way each component/hook decides:

```ts
import { crazyMethod } from 'api/madService'
```

**Comment by W3stside** [comment]
Interestingly, the functional API approach is how Dima and I coded the
first projects. (We both read YDKJS haha)

**Comment by anxolin** [comment]
> Interestingly, the functional API approach is how Dima and I coded the
first projects. (We both read YDKJS haha)

This means you also prefer to do new APIs as functions? we were coding most of things functionally, the APIs were done like this just because the idea was to have mock/impl and respecting the same interface, and be able to switch one for the other. 
I still think this is useful, but some API got bigger and more complex and it feels strange that's the only place we use OO.

**Comment by W3stside** [comment]
> > Interestingly, the functional API approach is how Dima and I coded the
> > first projects. (We both read YDKJS haha)
> 
> This means you also prefer to do new APIs as functions? we were coding most of things functionally, the APIs were done like this just because the idea was to have mock/impl and respecting the same interface, and be able to switch one for the other.
> I still think this is useful, but some API got bigger and more complex and it feels strange that's the only place we use OO.

Yes I think the classless approach using singletons is the way forward

**Review by W3stside — APPROVED** [review]

**Pull request merged**